src/plots/cartesian/layout_attributes.js

```diff
@@ -465,6 +465,36 @@ module.exports = {
             'Only has an effect if `anchor` is set to *free*.'
         ].join(' ')
     },
+    categorymode: {
+        valType: 'enumerated',
+        values: [
+            'trace', 'category ascending', 'category descending', 'array'
+            /*, 'value ascending', 'value descending'*/ // value ascending / descending to be implemented later
+        ],
+        dflt: 'trace',
+        role: 'info',
+        description: [
+            'Specifies the ordering logic for the case of categorical variables.',
+            'By default, plotly uses *trace*, which specifies the order that is present in the data supplied.',
+            'Set `categorymode` to *category ascending* or *category descending* if order should be determined by',
+            'the alphanumerical order of the category names.',
+            /*'Set `categorymode` to *value ascending* or *value descending* if order should be determined by the',
+            'numerical order of the values.',*/ // // value ascending / descending to be implemented later
+            'Set `categorymode` to *array* to derive the ordering from the attribute `categorylist`. If a category',
+            'is not found in the `categorylist` array, the sorting behavior for that attribute will be identical to',
+            'the *trace* mode. The unspecified categories will follow the categories in `categorylist`.'
+        ].join(' ')
+    },
+    categorylist: {
```

Comment:
`categories` implies strings. This should work for dates and numbers too, right?

We don't want to confuse `categories` with `ticklabels` (intuitively, I would think that `categories` would refer to the the tick labels or even the `x` data itself)

Other options:
- `tickorderarray` or `tickordervals` (if we use `tickorder` above)
-  `orderingvals`